### PR TITLE
(feature) Telegram bot

### PR DESCRIPTION
**Pull request by rhaxma**
Ticket: https://github.com/Swetrix/roadmap/issues/11

### CODE REVIEW

File: src/project/dto/project.dto.ts [hunk 1]
```diff
@@ -1,6 +1,25 @@
 import { ApiProperty } from '@nestjs/swagger'
 import { IsNotEmpty } from 'class-validator'
 
+export enum QueryMetric {
+  PAGE_VIEWS = 'page_views',
+  UNIQUE_PAGE_VIEWS = 'unique_page_views',
+}
+
+export enum QueryCondition {
+  GREATER_THAN = 'greater_than',
```
Comment (Blaumaus) [inline src/project/dto/project.dto.ts:10]: Can you also add GREATER_EQUAL_THAN (>=) and LESS_EQUAL_THAN (<=)?

File: src/analytics/analytics.service.ts [hunk 2]
```diff
@@ -807,4 +809,24 @@ export class AnalyticsService {
...
+    const query = `SELECT sid, dv, br, os, cc FROM analytics WHERE sid IN (${sids
+      .map(el => `'${el}'`)
+      .join(',')})`
+    const result = await clickhouse.query(query).toPromise()
+    const processed = _map(_uniqBy(result, 'sid'), el =>
+      _pick(el, ['dv', 'br', 'os', 'cc']),
+    )
+
+    return processed
+  }
```
Comment (Blaumaus) [inline src/analytics/analytics.service.ts:831]: you can use this instead:
```suggestion
    return await redis.countKeysByPattern(`hb:${pid}:*`)
  }
```

File: src/project/dto/project.dto.ts (hunk 1 above)
Comment (rhaxma) [reply src/project/dto/project.dto.ts:10]: Yes, of course.

File: src/auth/auth.controller.ts [hunk 3]
```diff
@@ -124,6 +131,40 @@ export class AuthController {
 
       user.sharedProjects = sharedProjects
 
+      const ua = UAParser(userAgent)
```
Comment (Blaumaus) [inline src/auth/auth.controller.ts:134]: this:
```javascript
      const ua = UAParser(userAgent)
      const br = ua.browser.name || 'unknown'
      const dv = ua.device.type || 'desktop'
      const os = ua.os.name || 'unknown'
      const tz = Intl.DateTimeFormat().resolvedOptions().timeZone
      const cc =
        ct.getCountryForTimezone(tz)?.id ||
        (headers['cf-ipcountry'] === 'XX' ? 'NULL' : headers['cf-ipcountry'])
      const country = ct.getCountry(cc)?.name || 'unknown'
```

should be put into the `if (user.isTelegramChatIdConfirmed) {` code block.

File: src/tg-integration/tg.update.ts [hunk 4]
```diff
@@ -1,27 +1,222 @@
...
+        `Online users: \`${onlineUserCount.length}\`` +
+        '\n' +
+        `Page views: \`${stats[project.id].thisWeek}\`` +
+        '\n' +
+        `Unique page views: \`${stats[project.id].thisWeekUnique}\`` +
+        '\n\n' +
+        '*Alerts*' +
+        '\n' +
+        `Online users: \`${
+          project.alertIfOnlineUsersExceeds ? 'seted' : 'not seted'
```
Comment (Blaumaus) [inline src/tg-integration/tg.update.ts:138]:
```suggestion
          project.alertIfOnlineUsersExceeds ? 'set' : 'not set'
```

File: src/tg-integration/tg.update.ts [hunk 5]
```diff
@@ -1,27 +1,222 @@
...
+        }\`` +
+        '\n' +
+        `Additional: \`${
+          project.isAdditionalAlertEnabled ? 'seted' : 'not seted'
```
Comment (Blaumaus) [inline src/tg-integration/tg.update.ts:142]:
```suggestion
          project.isAdditionalAlertEnabled ? 'set' : 'not set'
```

**Review by Blaumaus — CHANGES_REQUESTED** [review]

File: src/analytics/analytics.service.ts (hunk 2 above)
Comment (rhaxma) [reply src/analytics/analytics.service.ts:831]: No, because this function does not exist.

Comment (Blaumaus) [reply src/analytics/analytics.service.ts]: @rhaxma fyi it does https://github.com/Swetrix/swetrix-api/blob/main/src/common/constants.ts#L17

Comment (Blaumaus) [reply src/analytics/analytics.service.ts]: fixed it at https://github.com/Swetrix/swetrix-api/pull/92/commits/ea583c23666972abdd39e64b5ffb15059f7318c9

**Review by Blaumaus — APPROVED** [review]

**Pull request merged**